netrc: Implement .netrc parsing

davidpolverari opened the pull request:
This implementation parses the $HOME/.netrc (or a file specified by the
user) through the `--netrc[=file]` or `-R[file]` options.

It uses a simple FSM in order to parse the file and find the credentials
corresponding to the FTP host, or uses the "default" entry otherwise,
if any, according to the specification found on [1]. Tokens not
applicable to axel usage were not considered.

[1] https://docs.oracle.com/cd/E19455-01/806-0633/6j9vn6q5f/index.html

Closes: #116

Signed-off-by: David Polverari <david.polverari@gmail.com>

davidpolverari committed:
netrc: Implement .netrc parsing

This implementation parses the $HOME/.netrc (or a file specified by the
user) through the `--netrc[=file]` or `-R[file]` options.

It uses a simple FSM in order to parse the file and find the credentials
corresponding to the FTP host, or uses the "default" entry otherwise,
if any, according to the specification found on [1]. Tokens not
applicable to axel usage were not considered.

[1] https://docs.oracle.com/cd/E19455-01/806-0633/6j9vn6q5f/index.html

Closes: #116

Signed-off-by: David Polverari <david.polverari@gmail.com>

davidpolverari commented:
PS: On this implementation, I had to use the `// fall through` comment in order to placate GCC's `-Wimplicit-fallthrough`, as the project Makefile WARN_CFLAGS are considering a lot of warnings as errors and in this case it was the desired outcome.

As GCC considers this specific comment as an "explicit fallthrough", the error doesn't show, but I don't know whether other compilers (like clang) will present the same behavior.

ismaell commented on src/text.c:
```diff
@@ -161,6 +162,15 @@ main(int argc, char *argv[])
 				}
 			}
 			break;
+		case 'R':
```
Let's define a negative version, it's necessary for if there's a default in the axel config file. I'm not sure a short version is needed (I've been thinking about removing support for systems not supporting getopt_long anyway).

davidpolverari replied:
About negative version of the option, I don't know whether it is a good idea to make .netrc parsing default, as it may break existing users' assumptions/scripts/workflows, etc. What do you think about it?

ismaell replied:
I don't mean to make it the default, but if you have the option on the configuration file, then you need a way to disable it from the command line.

davidpolverari replied:
Ok, I'll add it.

ismaell commented on src/conf.h:
```diff
@@ -43,6 +43,7 @@
 #define AXEL_CONF_H
 
 typedef struct {
+	char netrc_filename[MAX_STRING];
```
Perhaps we can add an abstraction layer here, instead of saving the name, we can save a pointer to a struct where we can later add some form of caching at least. It would also be interesting to just keep a file descriptor instead of the file name.

davidpolverari replied:
I didn't understand what you meant in this case. Could you please elaborate on this point?

ismaell replied:
Add a struct pointer there, and add everything the parser needs inside it.

davidpolverari replied:
Ok.

ismaell commented on src/text.c:
```diff
@@ -161,6 +162,15 @@ main(int argc, char *argv[])
+		case 'R':
+			conf->use_netrc = 1;
+			if (optarg) {
+				if (!sscanf(optarg, "%s", conf->netrc_filename)) {
```
why `sscanf`?

davidpolverari replied:
`sscanf` was used to keep consistency with surrounding code.

ismaell replied:
`strlcpy` is better.

davidpolverari replied:
I was afraid of using strlcpy because it is not standard (neither ISO C nor POSIX), and thus could cause portability problems. But if it is ok for you, I'll use it.

davidpolverari replied:
PS: I couldn't use `strlcpy` under Linux, it is not available in libc. I'm afraid that to use that we'd have to implement it or include another lib on the project. I can use the approach I used elsewhere to emulate `strlcpy`: by using `strncpy` and null-terminating it.

ismaell reviewed (changes requested):
- whatever the parser needs should be into it's own struct, and it should get a pointer to it.
- allocate buffers with malloc/calloc to fail gracefully when OOM
- the state machine could be encoded as functions using tail merging
- don't use variables to stop a loop
- the parser shouldn't know anything about `axel_t`, outputs should be explicit too, pass pointers and lengths, so use `strlcpy`
- have you thought about using `mmap`? that could simplify the design.

davidpolverari commented:
Thanks for the review! I will try to address the points you made as soon as I can.

davidpolverari commented:
>     * have you thought about using `mmap`? that could simplify the design.

About this point, I was thinking about using `mmap`, but using it precludes the use of `strtok*`, because `mmap` won't null-terminate the mmaped file and `strtok`{,_r} will try to change the string, something unacceptable for a mmaped file, and something that will not work on read-only mappings, which would be the right way to mmap `.netrc` anyway.

I could either `malloc`(file_size + spare bytes), load file contents on this mem and null-terminate it, in order to work on that copy with strtok*, but in the unlikely event that someone has an absurdly big .netrc file, this approach might fail. Performance-wise, I don't think it would be a burden in most cases.

Or maybe I can just write a tokenizer function by iterating through the mmaped bytes and filling a `char []` with each found token, while keeping track of the beginning of the next token by means of local static pointer.

For me, the current `fgets` approach and the `malloc` one seem the most KISS. What are your views about it?

ismaell commented:
On 18/Oct/2019 20:51, davidpolverari wrote:
<...>
> Or maybe I can just write a tokenizer function by iterating through
> the mmaped bytes and filling a `char []` with each found token, while
> keeping track of the beginning of the next token by means of local
> static pointer.
> 
> For me, the current `fgets` approach and the `malloc` one seem the
> most KISS. What are your views about it?

It would be worth knowing how it compares in practice, I suspect it's
less overhead in the end, but for now the stdio API is fine.

It doesn't need to copy anything, and since the format is so simple, you
can get away with something as simple as strspn + strncmp, altough we
would need a limiting version of strspn, but that's about it.

davidpolverari committed:
WIP refactoring

davidpolverari commented on src/netrc.c:
```diff
@@ -0,0 +1,114 @@
+				}
+				break;
+			case LOGIN:
+				strcpy(conn->user, tok);
```
Those points are being addressed on the WIP implementation.

davidpolverari commented on src/netrc.c:
```diff
@@ -0,0 +1,114 @@
+	HOST_ENTRY,
+	LOGIN,
+	PASSWORD
+};
```
On the new implementation, states are encoded as functions.

davidpolverari commented on src/netrc.c:
```diff
@@ -0,0 +1,114 @@
+	char const *delims = " \t\n";
+	enum lookup_state state = NOTHING;
+
+	if (conn->conf->netrc_filename[0] == '\0') {
```
The support for the NETRC env var was added too.

davidpolverari commented:
I sent a new commit for the current PR. As soon as the code is ready for merging, I'll rebase everything.

davidpolverari commented on src/netrc.c:
```diff
@@ -37,78 +37,127 @@
+}
+
+int
+netrc_parse(const char *filename, const char *host, char *user, char *password)
```
I'm thinking about passing the length of `char *user` and `char *password` on this function interface, in order to be able to use them in `get_creds()` and make a safe `str{l,n}cpy` back to them from get_creds(), which will allow passing `char []` or a malloc'ed `char *` as those parameters from calling code.

ismaell replied:
Yes, you should pass pointer + length pairs.

davidpolverari replied:
Ok.

davidpolverari commented on src/netrc.c:
```diff
@@ -37,78 +37,127 @@
-		snprintf(conn->conf->netrc_filename, MAX_STRING, "%s/.netrc", home);
+
+	ret = 0;
+	if (filename && filename[0] != '\0') {
```
It is redundant when a `char *` is passed as filename, but not when a `char []` is passed, as it will be treated as valid (non-`NULL`) pointer, which it's might happen depending on the caller.

ismaell replied:
`[]` and `*` are exactly the same. Also it may look better this way: `if (filename && *filename) {...`

davidpolverari replied:
Yes, but a `char [] ` may be empty but its pointer will not be NULL. So if (filename) alone isn't enough. But yes, if (filename && *filename) looks better. I'll change it.

ismaell replied:
I was referring to expressions like `!!(x[0])` are equivalent to `!!(x[0] != '\0')`.

davidpolverari commented on src/netrc.c:
```diff
@@ -37,78 +37,127 @@
+	return st.st_size;
+}
+
+static int netrc_open(const char * const filename)
```
It was a mistake. I forgot to remove.

davidpolverari commented on src/netrc.c:
```diff
@@ -37,78 +37,127 @@
+		path = strdup(filename);
+	} else {
+		if ((netrc = getenv("NETRC"))) {
+			path = strdup(netrc);
```
Yes, that could be used, but we would have another special case to deal with, because the pointers returned by `getenv` wouldn't (and couldn't) be `free`d, but we'd have yet to `free` the allocated memory where we build the full path for the `.netrc` file.

I used strdup in this case to avoid dealing with those special cases.

davidpolverari replied:
Ok, understood. I already made the changes to avoid `strdup`.

davidpolverari commented on src/netrc.c:
```diff
@@ -37,78 +37,127 @@
+		e_addr = s_addr + sz;
+		ret = 1;
+	}
+	free(path);
```
We do not manage storage for environment variables. `free`ing a pointer returned by `getenv` is UB.

davidpolverari replied:
But I see your point. I refactored this and now I'm able to free the path right after opening it, but still using `strdup`.

ismaell replied:
Look at the snippet, I gave you exactly what you need to do to avoid the `strdup` for environment variables, you just need to keep an extra pointer.

davidpolverari replied:
I see now. I already changed to avoid `strdup` in those cases now.

davidpolverari commented on src/netrc.c:
```diff
@@ -37,78 +37,127 @@
+	return ret;
+}
+
+static void netrc_close(void)
```
Ok, I'll just `munmap` by the end of the parsing then.

ismaell replied:
Leave it mapped, just do nothing about it, we will reuse that.

davidpolverari commented on src/netrc.c:
```diff
@@ -37,78 +37,127 @@
+		if ((netrc = getenv("NETRC"))) {
+			path = strdup(netrc);
+		} else if ((home = getenv("HOME"))) {
+			size_t i = strlen(home) + strlen(REL_NETRC) + 1;
```
Ok.

davidpolverari commented on src/netrc.c:
```diff
@@ -37,78 +37,127 @@
+	while (IS_DELIM(p) && p < e_addr)
+		p++;
+	q = p;
+	while (!IS_DELIM(q) && q < e_addr)
```
Wouldn't it make the function too generic? I mean, we already know the tokens in `.netrc` file are separated only by spaces, tabs or new-lines, which are checked by the macro `IS_DELIM(x)`. I'm failing to see the benefits of the changes in case of this static aux function.

ismaell replied:
Well, every new functionality will likely influence refactoring of other parts of the program, so better be generic, it's not going to hurt footprint nor performance (in fact `strcspn` is generally faster than your `IS_DELIM` macro, believe it or not...).

ismaell replied:
Also, if you use some interface that looks like the standard one, chances are you'll be able to take that code and modify it with much less work.

davidpolverari commented on src/netrc.c:
```diff
@@ -37,78 +37,127 @@
+		} else if ((home = getenv("HOME"))) {
+			size_t i = strlen(home) + strlen(REL_NETRC) + 1;
+			path = malloc(i);
+			snprintf(path, i, "%s%s", home, REL_NETRC);
```
By using `strcpy` + `strcat` you mean?

ismaell replied:
Sort of, you need the size anyway, so you can do `memcpy` instead, on both.

```
size_t i = strlen(home);
path = malloc(i + sizeof(REL_NETRC))
memcpy(path, home, i);
memcpy(path+i, REL_NETRC, sizeof(REL_NETRC));
```

davidpolverari committed:
Latest WIP commit. Includes most of the requested changes.

Missing:

- add negative version (--no-netrc)

- add struct for netrc use, and use a pointer to it instead of using netrc_filename

davidpolverari commented:
I already addressed most of the points you made on my local repo. As soon as I implement the remaining ones (`netrc_t`, `strspn`-like interace, and `--no-netrc`) I will send the PR.

ismaell commented:
On 22/Oct/2019 18:15, davidpolverari wrote:
> I was afraid of using strlcpy because it is not standard (neither ISO
> C nor POSIX), and thus could cause portability problems. But if it is
> ok for you, I'll use it.

We provide strlcpy and strlcat for systems that don't have it on libc
and don't have libbsd.

It's OK to use non-standard functions if they're generic.

davidpolverari commented:
I didn't forget about this issue. I was busy because of work, but will resume as soon as possible.

davidpolverari commented:
I just sent another WIP commit with most of the proposed changes, including using `strspn` and `strcspn`-like functions (which I named `memspn` and `memcspn` as they don't deal with null-terminated strings).

Later on I will try to refactor `next_token()` into something like a `memtok_r()`, and will try to get rid of the global variables by explicitly passing them as arguments to the auxiliary static functions, besides implementing the remaining proposed changes.

davidpolverari commented on src/netrc.c:
```diff
@@ -41,75 +41,95 @@
+	char *aux = NULL;
+	char *home = NULL;
+	char *path = NULL;
+	const char suffix[] = "/.netrc";
```
I know, I know... But I thought this would be better than a `#define`. And if I used a `char *` instead of `char []`, that would imply usage of `strlen()` or hard-coded values.

davidpolverari committed:
Replace next_token with memtok and eliminate global vars.

davidpolverari committed:
Add netrc_t struct

davidpolverari committed:
Use netrc_t (wrongly) to test

davidpolverari commented:
I believe that with those latest PRs most of the requested things were changed. I have doubts now as to where we should keep the `netrc_t` pointer, at what point we should init the structure, etc. I think it would be easier if we could  decouple the username and password from `conn_t`, and had a struct used globally for it, but it would imply touching on a lot of places.

After that, I will be able to implement the negative (`--no-netrc`) option.

davidpolverari committed:
Refactor netrc module and its usage

davidpolverari commented:
Well, with this last PR I think the changes are almost all done, the only missing piece being the negative option (and removing the debugging `printf` I put). There are other things that I thought of doing at a later point in time, as moving the `memspn`, `memcspn` and `memtok` functions elsewhere on axel code, but I think that can be postponed.

Any thoughts about it? When everything is on right shape I can rebase it on master and push it again in a more organized way.

ismaell commented on src/netrc.c:
```diff
@@ -143,69 +144,59 @@ netrc_mmap(const char *filename, char **addr)
 	netrc = calloc(1, sizeof(netrc_t));
 	if (!netrc)
 		return NULL;
 	netrc->file = file;
```
Maybe here mmap the file somewhere and store only that pointer.

ismaell reviewed (changes requested):
Looks much better now. I made a couple of comments, the only important task now is to complete the get_creds implementation.

davidpolverari commented on src/conn.c:
```diff
@@ -254,12 +254,10 @@ conn_init(conn_t *conn)
-			netrc_t * netrc;
-			netrc = netrc_init(conn->conf->netrc_filename, conn->host, conn->user, sizeof(conn->user), conn->pass, sizeof(conn->pass));
-			netrc_parse(netrc);
+		if (conn->conf->netrc) {
```
Well, I will move this code to the `conn.c:conn_set()` function, as that is the one that deals with parsing URLs and extracting fields like username, password and hostname from them and copying to `conn_t`. But once parsing is entangled with auth code, I'd rather leave this kind of refactoring to another PR.

davidpolverari commented on src/text.c:
```diff
@@ -163,13 +163,12 @@ main(int argc, char *argv[])
+				if (optarg) {
+					netrc_filename = optarg;
 				}
+				conf->netrc = netrc_init(netrc_filename);
```
I left it that way because we if we pass `NULL` as the filename (when using the `--netrc` option and _not_ providing a filename), the `netrc.c:netrc_mmap()` function will first check whether the `NETRC` env var exists and will use its content as a the filename. Failing that, it will try to open `$HOME/.netrc`. Thus, I created this variable in order to avoid duplicating the `netrc_init()` call on an `else`. I'll try to refactor it without the var.

davidpolverari replied:
Done. Solved it with ternary operator.

davidpolverari committed:
conn.c: Move netrc credential parsing code to conn_set

davidpolverari committed:
text.c: Remove netrc_filename var

davidpolverari committed:
netrc.c: Break long line

davidpolverari committed:
text.c: remove useless ternary operator

davidpolverari committed:
netrc: Store pointer to mmaped file instead of filename

davidpolverari commented:
> Looks much better now. I made a couple of comments, the only important task now is to complete the get_creds implementation.

What else do you think it is missing on `get_creds()` implementation?

davidpolverari committed:
netrc: Hide netrc_t implementation details from header file

ismaell committed:
WIP netrc support

- removed netrc_free
- simplified netrc_mmap
- reworked parser

ismaell commented:
@davidpolverari please take a look at https://github.com/ismaell/axel/commit/888ec1e6f8356d7ac3a768be77089a72bad12b7b.

That fixes both the null deref at the mmap function, the parsing (which now aborts on unkonwn tokens), and removes two globals.

What I'm missing there is a cache, so that each machine is only parsed once.

davidpolverari committed:
WIP: netrc support

- simplified parser flow
- replaced memcpy with strlcpy

davidpolverari commented:
Your parser implementation is way more elegant than mine :smile:. Anyways, I had to make minor changes to make it work and to simplify (IMHO) the program flow. Besides, we can't use `memcpy` because the pointers returned by `memtok` aren't NULL-terminated, thus we need `strlcpy` to NULL-terminate it the copy for us.

As the parser now aborts on unknown tokens, we may need to include the other tokens used by the format which are unused on axel, and turn them in no-ops, otherwise a compliant file might get ignored.

davidpolverari commented:
Regarding the caching (memoization), I think we could create a function to wrap `netrc_parse` and populate/verify/return the previous results contained in an array of few elements of a struct (host, username, password) inside `netrc_t`, acting like a circular buffer. With few elements, we wouldn't need any fancy algorithm to do it.

But don't you think it is better to implement it as a separate PR later?

ismaell commented on src/netrc.c:
```diff
@@ -198,19 +199,24 @@ netrc_parse(netrc_t *netrc, const char *host, char *user, size_t user_len, char
 				if (!strncmp(host, tok.data, tok.len))
 					matched = true;
-			} else
+			} else if (tok.data[0] == 'd') {
```
I would leave this line out, since there's only two options.

ismaell commented on src/netrc.c:
```diff
@@ -198,19 +199,24 @@ netrc_parse(netrc_t *netrc, const char *host, char *user, size_t user_len, char
+				}
+
+				/* need strlcpy here, as tok.data isn't NULL-terminated */
+				strlcpy(p->dst, tok.data, tok.len + 1);
```
I would still prefer memcpy here, as we know exactly where the zero should go.

ismaell committed:
WIP netrc: improve netrc_parse token copying

ismaell committed:
WIP netrc: make parser struct R/O and make netrc_parse void

ismaell committed:
WIP netrc: move null check to netrc_parse

ismaell commented on src/netrc.c:
```diff
@@ -0,0 +1,222 @@
+	return NULL;
+}
+
+int
```
Probably it should be made void, since the return value isn't being used.

davidpolverari replied:
We don't use it ATM, but we might need it depending on how we decide to deal with failure later.

ismaell replied:
Then we should change that later, when we actually need it. The problem is sometimes that day never comes.

davidpolverari replied:
Well, change was made and it is now `void`, but I didn't see any impact it returning `int` would have. I think it wouldn't hurt and it could be used both ways (worrying or not about the return code). It would be up to the calling code to decide on policy, and there would be no need to change `netrc_parse` to accommodate them.

ismaell commented on src/netrc.c:
```diff
@@ -0,0 +1,222 @@
+
+int
+netrc_parse(netrc_t *netrc, const char *host, char *user, size_t user_len, char *pass, size_t pass_len)
+{
```
The check for `NULL == netrc` should be done here instead of at the caller.

davidpolverari replied:
The check made by the caller is meant to check if `--netrc` was used (`netrc_init` will only init the struct in this case), so I think we shouldn't remove that. But I'll add a check here anyway.

ismaell replied:
It should be called unconditionally, it's almost always a better strategy, and it allows to implement other mechanisms while avoiding boilerplate code at the callers at the same time.

For example, this could be used to feed the default password for FTP instead of hard-coding it.

davidpolverari replied:
Now I see your point on this. It is better indeed.

ismaell commented on src/netrc.c:
```diff
@@ -0,0 +1,222 @@
+{
+	bool matched = false;
+	buffer_t tok, save_buf = {};
+	struct parser {
```
should be `const struct`

davidpolverari replied:
Ok.

davidpolverari committed:
Minor changes

ismaell commented:
I failed to mention I made changes at https://github.com/ismaell/axel netrc branch

davidpolverari committed:
Revert "Minor changes"

This reverts commit c2fc839a40e1ebe1282e481963ceb433db981f9f.

davidpolverari commented:
I cherry-picked your commits here, as they would be the same changes I'd have to make.

davidpolverari commented:
Hi! Yes, was just a little bit busy with other stuff.

ismaell commented:
This was merged, with some changes in commit 216a8e08e7c709d8218fffb08dcb671423b0cdd3.
